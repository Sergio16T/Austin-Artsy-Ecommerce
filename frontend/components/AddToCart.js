import React, { Component } from 'react';
import { Mutation } from 'react-apollo'; 
import gql from 'graphql-tag'; 
import { CURRENT_USER_QUERY } from './User'; 

const ADD_TO_CART_MUTATION = gql`
    mutation ADD_TO_CART_MUTATION($id: ID! ) {
        addToCart(id: $id) {
            id
            quantity
        }
    }
`; 

class AddToCart extends Component {
    render() {
        return (
            <Mutation 
            mutation={ADD_TO_CART_MUTATION} 
            variables={{id: this.props.id}}
            refetchQueries={[{query: CURRENT_USER_QUERY }]}
            >
                {(addToCart, {error, loading}) => 
                    <button type="button" onClick={async () => {
                        const res = await addToCart().catch(err => {
                            alert(err.message); 
                        }); 
                        console.log(res);
                    }}>Add to Cart</button>
                }
            </Mutation>
        );
    }
}

export default AddToCart;