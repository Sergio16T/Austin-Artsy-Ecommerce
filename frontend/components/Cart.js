import React, { Component } from 'react';
import styled from 'styled-components';

const CartWrapper = styled.div`
    position: fixed; 
    height: 100%; 
    width: 400px; 
    transform: ${props => props.isOpen ? "translateX(0)" : "translateX(100%)" }; 
    background: ${props => props.theme.offWhite}; 
    right: 0; 
    top: 0; 
    z-index: 5; 
    border-left: 2px solid ${props => props.theme.purple}; 
    min-width: 500px; 
    transition: all 0.3s;
    overflow: hidden; 
`; 

class Cart extends Component {
    render() {
        return (
            <CartWrapper isOpen={this.props.isOpen}>
                <h2>Cart</h2>
            </CartWrapper>
        );
    }
}

export default Cart;