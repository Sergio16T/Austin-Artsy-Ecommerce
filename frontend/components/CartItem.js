import React from 'react';
import { useMutation } from 'react-apollo'; 
import gql from 'graphql-tag'; 
import styled from 'styled-components';
import RemoveFromCart from './RemoveFromCart'; 
import formatMoney from "../lib/formatMoney";
import { CURRENT_USER_QUERY } from './User';

const UPDATE_CART_ITEM_MUTATION = gql`
    mutation UPDATE_CART_ITEM_MUTATION($updateType: String!, $id: ID!) {
        updateCartItem(updateType: $updateType, id: $id) {
            id
            quantity
        }
    }
`; 

const StyledCartItem = styled.div`
display: grid; 
grid-template-columns: 1fr 2fr .5fr; 
padding: 2rem 0; 
.cartItemImage {
    width: 120px; 
}
@media (max-width: 600px) {
    padding: 1rem 0; 
    .cartItemImage {
        width: 80px; 
    }
}
`; 

const CartItemDescription = styled.div`
    padding-left: 2rem;
.cartitem_title{
    margin-top: 0; 
    font-size: 1.2rem;
}
.priceandquantity {
    display: flex; 
    flex-direction: row; 
}
`; 
const QuantityDiv = styled.div`
    display: flex;
    flex-direction: row; 
    font-size: 1.6rem; 
    padding-left: 2rem; 
    align-items: center; 
.cartItem_Quantity, .minus, .plus {
    padding:  0 .75rem; 
    margin: 0; 
}
.minus, .plus {
    outline: none; 
    border: none; 
    background: transparent; 
    font-size: 1.6rem; 
    cursor: pointer; 
}
.cartitem_price {
    font-size: 1.2rem; 
}
`; 
const CartItem = (props) => {
    const [updateCartItem, { data }] = useMutation(UPDATE_CART_ITEM_MUTATION); 
	const { item } = props.cartItem; 
	return (
		<StyledCartItem>
			<img className="cartItemImage" src={item.image}/> 
			<CartItemDescription>
				<p className="cartitem_title">{item.title}</p>
				<div className="priceandquantity">
					<p className="cartitem_price">{formatMoney(item.price)}</p>
					<QuantityDiv>
						<button className="minus" type="button" onClick={() => {
                            updateCartItem({variables: { updateType: "subtract", id: props.cartItem.id}}); 
                        }} disabled={props.cartItem.quantity === 1}>-</button>
						<p className="cartItem_Quantity">{props.cartItem.quantity}</p>
						<button className="plus" onClick={() => {
                            updateCartItem({variables: { updateType: "add", id: props.cartItem.id}}); 
                        }}>+</button>
					</QuantityDiv>
				</div>
			</CartItemDescription>
			
			<RemoveFromCart id={props.cartItem.id}/>
		</StyledCartItem>
	)
}

export default CartItem; 