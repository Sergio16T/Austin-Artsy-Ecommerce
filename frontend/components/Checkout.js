import React, { Component } from 'react';
import StripeCheckout from 'react-stripe-checkout'; 
import { Mutation } from 'react-apollo'; 
import Router from 'next/router'; 
import Nprogress from 'nprogress'; 
import PropTypes from 'prop-types'; 
import gql from 'graphql-tag'; 
import calculateTotal from '../lib/calculateTotal'; 
import { CURRENT_USER_QUERY } from './User'; 

const CREATE_ORDER_MUTATION = gql`
    mutation CREATE_ORDER_MUTATION($token: String!) {
        createOrder(token: $token) {
            id
						total 
						charge
						items {
							id
							title
						}
        }
    }
`; 

function totalItems(cart) {
    return cart.reduce((startValue, element) => {
        return startValue + element.quantity; 
    }, 0); 
}

class Checkout extends Component {
    onToken = async (res, createOrder) => {
				console.log(res.id); 
				this.props.toggleCart(false);
				const order = await createOrder({
					variables: { token: res.id }
				}).catch(err => {
					console.log(err.message);  
				}); 
				console.log(order); 
				Router.push({
					pathname: "/order", 
					query: {id: order.data.createOrder.id}
				});
			
    }
    render() {
        return (
						<Mutation 
						mutation={CREATE_ORDER_MUTATION}
						refetchQueries={[{ query: CURRENT_USER_QUERY }]}
						>
                {(createOrder, {error, loading}) => 
							<StripeCheckout
							amount={this.props.totalAmount}
							name="Austin's List"
							description={`Order of ${totalItems(this.props.cart)} items`}
							image={this.props.cart.length && this.props.cart[0].item ? this.props.cart[0].item.image : null}
							stripeKey="pk_test_lUHGmLxt0KGqdsIN20tuXiVw00Q2OwgjiR"
							currency="USD"
							email={this.props.email}
							token={(res) => this.onToken(res, createOrder)}
							>{this.props.children}</StripeCheckout>
            }
            </Mutation>
        );
    }
}

export default Checkout;