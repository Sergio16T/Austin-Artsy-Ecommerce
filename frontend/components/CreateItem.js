import React, { Component } from 'react';
import { Query, Mutation } from 'react-apollo'; 
import Router from 'next/router'; 
import gql from'graphql-tag'; 
import { StyledFormWrapper, StyledForm} from './styles/FormStyles'; 
import Spinner from './Spinner'; 
import { ALL_ITEMS_QUERY } from './Items'; 
import { PAGINATION_QUERY } from './Pagination'; 

const CREATE_ITEM_MUTATION = gql`
    mutation CREATE_ITEM_MUTATION(
        $title: String! 
        $description: String!
        $price: Int! 
        $image: String!
        $largeImage: String!
        ) {
        createItem(
            title: $title, 
            description: $description, 
            price: $price
            image: $image
            largeImage: $largeImage 
            ) {
            id
        }
    }
`; 
class CreateItem extends Component {
    state = {
        title: '', 
        price: 0, 
        description: '',
        image: '',
        largeImage: '',
        errorMessage: '',
        uploadError: ''
    }
    uploadFile = async (e) => {
        console.log('uploading file'); 
        this.setState({ spinner: true }); 
        const files = e.target.files; 
        const data = new FormData(); 
        data.append('file', files[0]); 
        data.append('upload_preset', 'AustinArts'); 

        const res = await fetch('https://api.cloudinary.com/v1_1/dddnhychw/image/upload', {
            method: 'POST', 
            body: data  
        }); 
        const file = await res.json(); 
        console.log(file); 

        try {
            this.setState({
                image: file.secure_url, 
                largeImage: file.eager[0].secure_url,
                spinner: false
            }); 
        }
        catch(err) {
            this.setState({uploadError: err.message}); 
        }
    }
    handleInput = (e) => {
        let regEx = /^\d+(\.\d{0,2})?$/; 
        const { value, type, name } = e.target; 
        let val = type === 'number' && value.length ? parseFloat(value) : value;  
        // console.log(regEx.test(val)); 
        try {
            if ((typeof val === 'number') && !regEx.test(val)) {
                throw new Error('Only two decimal points please'); 
            }
        }
        catch(err) {
            console.log(err.message);
            this.setState({
                errorMessage: err.message
            }); 
        }
        if ((typeof val === 'number') && regEx.test(val)) {
            this.setState({
                errorMessage: ''
            }); 
        }
        this.setState({
            [name]: val
        }); 
    }
    render() {
        let regEx2= /^\d+(\.\d{0,1})$/; 
        const { errorMessage } = this.state; 
        const setPrice = regEx2.test(this.state.price) ? this.state.price.toFixed(2) : this.state.price; 
        return (
            <StyledFormWrapper>
                <div className="formContainer">
                    <Query query={PAGINATION_QUERY}>
                        {({data, error, loading}) => {
                            if(loading) return null; 
                            const count = data.itemsConnection.aggregate.count; 
                            // console.log(count); 
                            let lastOnPage  = count/4 === 0; 
                            let page = Math.ceil((count+1)/4) 
                            // console.log(page); 
                            let skip = page * 4 - 4;
                            // console.log(skip); 
                            return (
                        <Mutation 
                        mutation={CREATE_ITEM_MUTATION} 
                        variables={{...this.state, price: setPrice * 100}}
                        refetchQueries={[
                            { query: PAGINATION_QUERY }, 
                            { query: ALL_ITEMS_QUERY, variables: {skip: skip} }, 
                            { query: ALL_ITEMS_QUERY, variables: {skip: (page-1) * 4 - 4 }}
                        ]}
                        >
                            {(createItem, {loading, error}) => (
                            <StyledForm onSubmit={ async (e)=> {
                                e.preventDefault(); 
                                this.setState({ spinner: true });
                                let res = await createItem().catch(err => {
                                    if(err) {
                                        this.setState({ 
                                            spinner: false, 
                                            error: true
                                         }); 
                                    } else {
                                        this.setState({
                                            error: false
                                        }); 
                                    }  
                                }); 
                                this.setState({ spinner: false }); 
                                console.log(res); 
                                if(!this.state.error) {
                                    Router.push({
                                        pathname: "/items", 
                                        query: {page: page}
                                        // query: {id: res.data.createItem.id}
                                    }); 
                                }
                            }} 
                            errorMessage = {errorMessage}
                            >
                                <Spinner spinner={this.state.spinner}/>
                                {error && <p className="errorMessage">{error.message.replace("GraphQL error:", "")}</p>}
                                <fieldset disabled={loading} aria-busy={loading}>
                                    <label htmlFor="file">
                                        Image
                                        <input
                                        type ="file" 
                                        id ="file" 
                                        name="file" 
                                        placeholder="upload an image" 
                                        onChange={this.uploadFile}
                                        />
                                    </label>
                                    {this.state.largeImage && <img width="250" src={this.state.image} alt={this.state.description}></img>}
                                    <label htmlFor="title">
                                        Title
                                        <input
                                        type ="text" 
                                        id ="title" 
                                        name="title" 
                                        placeholder="title" 
                                        value={this.state.title}
                                        onChange={this.handleInput}
                                        required  
                                        />
                                    </label>
                                    <label htmlFor="price" id="priceLabel">
                                        {errorMessage ? errorMessage : 'Price'}
                                        <input
                                        type ="number" 
                                        id ="price" 
                                        name="price" 
                                        placeholder="price" 
                                        value={this.state.price}
                                        step=".01"
                                        onChange={this.handleInput}
                                        required  
                                        />
                                    </label>
                                    <label htmlFor="description">
                                        Description
                                        <textarea
                                        id ="description" 
                                        name="description" 
                                        placeholder="Enter a Description" 
                                        value={this.state.description}
                                        onChange={this.handleInput}
                                        required  
                                        />
                                    </label>
                                    <button type="submit" disabled={errorMessage ? true : false} aria-disabled={errorMessage ? true : false}> Submit </button>
                                </fieldset>
                            </StyledForm>
                        )}
                        </Mutation> 
                    )}}
                </Query>
                </div>
            </StyledFormWrapper>
        );
    }
}

export default CreateItem;