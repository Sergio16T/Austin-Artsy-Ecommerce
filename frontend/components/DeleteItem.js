import React, { Component } from 'react';
import { Mutation } from 'react-apollo'; 
import gql from 'graphql-tag';
import Router from 'next/router'; 
import { ALL_ITEMS_QUERY } from './Items'; 
import { PAGINATION_QUERY } from './Pagination';

const DELETE_ITEM_MUTATION = gql`
    mutation DELETE_ITEM_MUTATION($id: ID!) {
        deleteItem(id: $id) {
            id
            title
        }
    }
`

let query = gql`
query ALL_ITEMS_QUERY($skip: Int = 0, $first: Int = 4) {
    items(skip: $skip, first: $first, orderBy: createdAt_ASC) {
        id
        title
        price
        description
        image 
        largeImage
    }
}
`; 

class DeleteItem extends Component {
    render() {
        return (
        <Mutation 
        mutation={DELETE_ITEM_MUTATION} 
        variables={{id: this.props.id}}
        refetchQueries={[
            {query: ALL_ITEMS_QUERY, variables:{skip: this.props.page * 4 -4}},
            {query: ALL_ITEMS_QUERY, variables:{skip: (this.props.page - 1) * 4 -4}},
            {query: ALL_ITEMS_QUERY, variables:{skip: (this.props.page + 1) * 4 -4}},
            {query: PAGINATION_QUERY },
        ]}
        >
            {(deleteItem, {error, loading}) => 
                <button onClick={async() => {
                    if(confirm("Are you sure you want to delete this item?")) { 
                    let res = await deleteItem().catch(err => {
                        alert(err.message); 
                    }); 
                    console.log('delete res', res); 
                    // Router.push({
                    //     pathname: "/items", 
                    //     query: {page: this.props.page}
                    // }); 
                    }
                }}>{this.props.children}</button>
            }
        </Mutation>
        
        );
    }
}

export default DeleteItem;