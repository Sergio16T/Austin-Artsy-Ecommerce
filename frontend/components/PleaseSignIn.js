import { Query } from 'react-apollo'; 
import SignIn from "./SignIn"; 
import { CURRENT_USER_QUERY } from './User'; 
import { StyledFormWrapper } from './styles/FormStyles'; 

const PleaseSignIn = props => {
    return (
        <Query query={CURRENT_USER_QUERY}>
            {({data, loading}) => {
                // console.log(data); 
                if (loading) return null; 
                if(!data.user)  { return (
                    <StyledFormWrapper data-testid="pleaseSignin">
                        <div className="formContainer">
                            <SignIn/>
                        </div>
                    </StyledFormWrapper>
                )}
                return props.children; 
            }}
        </Query>
    );
};



export default PleaseSignIn;