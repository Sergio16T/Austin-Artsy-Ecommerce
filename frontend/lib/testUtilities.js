import casual from 'casual'; 
//.seed() is to provide ability to use a specific seed in order to get a repeatable random sequence for our testing purposes
casual.seed(111); 

const mockItem = () => ({
    // __typename: 'Item',
    id: '123',
    price: 5000,
    user: null,
    image: 'dog.jpg',
    title: 'dogs are the best',
    description: 'dogs description',
    largeImage: 'dog-large.jpg',
}); 

const mockUserWithItemsInCart = () =>  ({
    __typename: 'User', 
    id: "123", 
    name: casual.name, 
    email: casual.email, 
    permissions: ["ADMIN"], 
    cart: [{
        id: 'abc', 
        quantity: 2, 
        item: {
            id: 'dgh', 
            price: 15000, 
            image: 'www.dog.image.com',
            title: "puppy"
        }
        },
    ]
}); 
const mockUser = () => ({
    __typename: 'User', 
    id: "123", 
    name: casual.name, 
    email: casual.email, 
    permissions: ['ADMIN'], 
    cart: [],
    orders: []
}); 
const mockOrderItem = () => ({
    __typename: 'OrderItem',
    id: casual.uuid,
    image: `${casual.word}.jpg`,
    title: casual.words(),
    price: 4234,
    quantity: 1,
    description: casual.words(),
  });
  
  const mockOrder = () => ({
    __typename: 'Order',
    id: 'ord123',
    charge: 'ch_123',
    total: 40000,
    items: [mockOrderItem(), mockOrderItem()],
    createdAt: '2018-04 - 06T19: 24: 16.000Z',
    user: mockUser(),
  });
  
  const mockCartItem = overrides => ({
    __typename: 'CartItem',
    id: 'omg123',
    quantity: 3,
    item: mockItem(),
    user: mockUser(),
    ...overrides,
  });
//mock LocalStorage 
class MockLocalStorage {
    constructor(){
        this.store={}; 
    }

    clear() {
        this.store = {}; 
    }
    getItem(key) {
        return this.store[key] || null; 
    }
    setItem(key, value) {
        this.store[key] = value.toString(); 
    }
    removeItem(key) {
        delete this.store[key]; 
    }
}

export { 
    mockItem, 
    MockLocalStorage, 
    mockUser, 
    mockCartItem, 
    mockOrderItem, 
    mockOrder }; 