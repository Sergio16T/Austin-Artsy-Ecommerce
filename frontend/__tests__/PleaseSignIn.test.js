import { mount } from 'enzyme'; 
import PleaseSignIn from '../components/PleaseSignIn'; 
import { CURRENT_USER_QUERY } from '../components/User'; 
import { MockedProvider } from '@apollo/react-testing'; 
import { mockUser} from '../lib/testUtilities'; 
import { act } from 'react-dom/test-utils'; 
import wait from 'waait';  

const notSignedInMocks = [{ 
        request: { query: CURRENT_USER_QUERY}, 
        result: { data: { user: null } }, 
    }, 
]; 

const signedInMocks = [{
    request: { query: CURRENT_USER_QUERY }, 
    result: { data: { user: mockUser() } }, 
}];

describe('<PleaseSignIn/>', () => {
    it('renders the sign in dialog to logged out users ', async () => {
        const wrapper = mount(
            <MockedProvider mocks={notSignedInMocks}>
                <PleaseSignIn/>
            </MockedProvider>
        ); 
        
        await wait(); 
        act( () => {
             wrapper.update(); 
        }); 
        console.log(wrapper.debug()); 
        expect(wrapper.text()).toContain("Sign In"); 
        expect(wrapper.find('SignIn').exists()).toBe(true); 
    }); 
    it('renders the child component when the user is signed in', async () => {
        const Child = () => <p>Child Component</p>; 
        const wrapper = mount(
            <MockedProvider mocks={signedInMocks}>
                <PleaseSignIn>
                    <Child/>
                </PleaseSignIn>
            </MockedProvider>
        ); 
        await wait(); 
        act( () => {
            wrapper.update(); 
       }); 
        console.log(wrapper.debug()); 
        expect(wrapper.find('Child').exists()).toBe(true); 
    });
})