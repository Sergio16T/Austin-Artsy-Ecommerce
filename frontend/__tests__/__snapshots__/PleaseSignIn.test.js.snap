// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PleaseSignIn/> renders the sign in component for logged out users 1`] = `
<div>
  <div
    class="FormStyles__StyledFormWrapper-txoayt-0 cmAGHL"
    data-testid="pleaseSignin"
  >
    <div
      class="formContainer"
    >
      <form
        class="FormStyles__StyledForm-txoayt-1 jOxosP"
        method="post"
      >
        <div
          class="Spinner__StyledSpinner-ma7w1f-0 VyxpA"
        >
          <div
            id="backDrop"
          >
            <div
              id="loading"
            />
          </div>
        </div>
        <fieldset
          aria-busy="false"
        >
          <h2>
            Sign in
          </h2>
          <label
            for="email"
          >
            Email
            <input
              name="email"
              placeholder="Enter your email address"
              type="text"
              value=""
            />
          </label>
          <label
            for="password"
          >
            Password
            <input
              name="password"
              placeholder="Password"
              required=""
              type="password"
              value=""
            />
          </label>
          <button
            type="submit"
          >
             Sign In 
          </button>
        </fieldset>
        <div
          class="FormStyles__StyledResetNote-txoayt-2 hTHisz"
        >
          <p
            id="resetNote"
          >
            Forgot your Password? Request reset
          </p>
        </div>
      </form>
    </div>
  </div>
</div>
`;
